### ISSUE TITLE
Fabricjs canvas inside Angular Material Tabs doesn't work

### ISSUE THREAD

**Issue by rak13** [kangax/fabric.js#2896]
I have been trying to use Fabric.canvas inside [angular material tabs](https://material.angularjs.org/1.0.7/demo/tabs) but that doesn't seem to work. Finally I want there to be multiple tabs with multiple canvases. For now I'm trying it with only one tab.

[This code without tabs seems to wok just fine, it allows free hand drawing with mouse](http://js.do/rak13/no-tabs)
But,
[this html code with angular material tabs doesn't let any drawing or event on the canvas](http://js.do/rak13/with-angular-material-tabs)

The javascript code (at the end) used in both the files are the same

How do I make it work?

**Comment by rak13** [comment]
@kangax please reply soon...

**Comment by isnifer** [comment]
@rak13 hey! I saw your code.
I think your problem with tabs init. Let see what's happen:
1. You create new fabric canvas
2. Angular tabs are created

And after that, your original canvas no more exists.
Just wrap your code with setTimeout:
```js
setTimeout(function () {
    var canvas =  new fabric.Canvas("myCanvas");
    console.log(canvas);
    canvas.setWidth(400);
    canvas.setHeight(400);
    canvas.isDrawingMode = true;
    console.log(canvas.isDrawingMode);
}, 0);
```
And it works!
![fabric_js](https://cloud.githubusercontent.com/assets/1788245/14582991/36a57f60-0426-11e6-9ce4-a437ac883a08.gif)

**Comment by rak13** [comment]
@isnifer thanks! 
that explained it all :+1: 
so, i need to wait for angular tabs to be loaded first and then initialize fabric canvas.

**Comment by asturur** [comment]
i do not know angular very good. look if there is a callback instead of using the setTimeout.

please update your stackoverflow question with the answer maybe will prevent other people to do same question here.

**Comment by isnifer** [comment]
@rak13 you're welcome!

**Comment by isnifer** [comment]
Yeah, callback would be better than setTimeout if it possible :)

**Comment by rak13** [comment]
how can i use callback?
and setTimeout(fun, 0), a timeout of 0 ms and it still works!!

**Comment by asturur** [comment]
the callback either is included in angular or is not possible to use it. read the docs.
i bet angular has some function that is called upon dom readyness.

the setTimeout 0 works probably because the time needed to angular to init the tabs is really nothing.

**Comment by isnifer** [comment]
Read the API docs. If it possible - use it, if not - use setTimeout

**Comment by shurui91** [comment]
I've read all the conversations. Just one thing I want to ask: is there anyway we can use Fabric.js in an Angular way (directive, service and such)? Or we have to load Fabric.js from index page (usually, since ui.route is there) and write non-Angular JS code in pages? I just want to see what is the better practice.

**Comment by asturur** [comment]
What do you mean by 'non angular code'?

**Comment by shurui91** [comment]
Sorry, let me rephrase. Yes, we could directly using Fabric.js without applying any Angular directives, services and etc in an Angular project. For example, when defining a Fabric canvas, we use <canvas> in HTML and Fabric.canvas('myCanvas'). But I am trying to see if we have something like a Angular directive, maybe like '<fabric-canvas>' tag available for angular project. I know we could directly insert Fabric js into Angular, but I am just looking for a more elegant way to work with Angular.

I guess this could also extend to be a discussion of "what is the best way we can use 3rd party libraries to work with AngularJS?". So far, if the libraries do not have its corresponding directives for Angular, I think we have to do it in an non-Angular way / pure JS way in order to work with Angular.
